design/DESIGN_typed_context_pipeline.md

```diff
@@ -0,0 +1,451 @@
+    // to the final TCurrentChainOutputView.
+    final TCurrentChainOutputView Function(TCurrentChainInputView)
+        completeMiddlewareChain = _chain;
+
+    return (NewContext ctx) {
+      // This cast assumes the pipeline was started with `PipelineBuilder.start()`,
+      // making TCurrentChainInputView effectively BaseContextView.
+      final initialView =
+          BaseContextView(ctx) as TCurrentChainInputView;
+
+      // Execute the middleware chain.
+      final finalView = completeMiddlewareChain(initialView);
+
+      // Execute the final handler with the processed context view.
+      return handler(finalView);
+    };
+  }
+}
+```
+
+## 5. Usage Example
+
+This demonstrates how to build a pipeline and how type errors would be caught.
+
+```dart
+// Assuming SessionMiddleware is:
+// UserSessionContextView sessionMiddleware(UserContextView userView) { ... }
+
+// Define a handler that expects both User and Session data.
+Future<Response> mySecureHandler(UserSessionContextView context) async {
+  final user = context.user;
+  final session = context.session;
+  return Response.ok('User: ${user.email}, Session ID: ${session.sessionId}');
+}
+
+void setupServer() {
+  final requestHandler = PipelineBuilder.start() // Starts with BaseContextView
+      .add(authenticationMiddleware)             // Output view: UserContextView
+      .add(sessionMiddleware)                    // Input: UserContextView, Output: UserSessionContextView
+      .build(mySecureHandler);                   // Handler expects UserSessionContextView - OK!
+
+  // This handler can now be passed to Relic's server serving mechanism.
+  // relicServe(requestHandler, ...);
+
+  // Example of a compile-time error:
+  // final faultyHandler = PipelineBuilder.start()
+  //     .add(authenticationMiddleware) // Output: UserContextView
+  //     // Missing sessionMiddleware
+  //     .build(mySecureHandler);      // COMPILE ERROR: mySecureHandler expects UserSessionContextView,
+  //                                  // but pipeline only guarantees UserContextView.
+}
+```
+
+## 6. Benefits
+
+*   **Compile-Time Safety**: The primary goal. Misconfigured pipelines (e.g., missing middleware, incorrect order affecting context data) are caught by the Dart analyzer.
+*   **Improved Developer Ergonomics**:
+    *   Handlers and middleware can declare precisely the context view (and thus data) they expect.
+    *   Access to context data via extension type getters is type-safe and clear (e.g., `context.user`).
+*   **Minimal Runtime Overhead for Views**: Extension types are intended to be zero-cost compile-time wrappers. The `ContextProperty` helper encapsulates `Expando` lookups/attachments, which are generally efficient.
+*   **No Class Explosion**: Avoids needing a distinct context `class` for every possible combination of middleware. Extension types provide views, and `ContextProperty` manages data association with the stable `requestToken`.
```

Comment:
> Avoids needing a distinct context `class` for every possible combination of middleware.

Is this true though? Maybe I misunderstand some aspect of the implementation or how I could avoid that, but I feel like I would always need to make the "next" middleware aware of the previous (and thus alter it's type when pre-pending a new one). (Assuming all middleware enhances the context with local data and is not just a passthrough like the logger in this example.)